components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
         }
     }
 
+    @Suppress("ComplexMethod")
+    internal fun createHistoryDelegate() = object : GeckoSession.HistoryDelegate {
+        @Suppress("TooGenericExceptionCaught")
+        override fun onVisited(
+            session: GeckoSession,
+            url: String,
+            lastVisitedURL: String?,
+            flags: Int
+        ): GeckoResult<Boolean>? {
+            if ((flags and GeckoSession.HistoryDelegate.VISIT_TOP_LEVEL) == 0 ||
+                (flags and GeckoSession.HistoryDelegate.VISIT_UNRECOVERABLE_ERROR) != 0) {
+
+                return GeckoResult.fromValue(false)
+            }
+
+            val isReload = lastVisitedURL?.let { it == url } ?: false
+            val result = GeckoResult<Boolean>()
+            scope.launch {
+                var exception: Throwable? = null
+                try {
+                    settings.historyTrackingDelegate?.onVisited(url, isReload, privateMode)
```

Comment:
We shouldn't even call into `historyTrackingDelegate` if we're in a `privateMode`.

Reply:
In that case, should we also remove the `privateMode` argument from `historyTrackingDelegate.{onVisited, getVisited}`?

Reply:
On one hand, `privateMode` is useful on a conceptual level in the delegate as it allows to easily segment history into two buckets, and let the storage layer do something sensible with those buckets. On another hand, we don't actually follow through on this, and it's not clear that we'll want to. I'll file a follow-up to consider cleaning this up.

Reply:
Ah! I see that you killed that argument. That's fine!